wsproto/connection.py

```diff
@@ -112,6 +112,41 @@ def receive_data(self, data):
 
         if self.state in (ConnectionState.OPEN, ConnectionState.LOCAL_CLOSING):
             self._proto.receive_bytes(data)
+            try:
+                for frame in self._proto.received_frames():
+                    if frame.opcode is Opcode.PING:
+                        assert frame.frame_finished and frame.message_finished
+                        self._events.append(Ping(payload=frame.payload))
+                    elif frame.opcode is Opcode.PONG:
+                        assert frame.frame_finished and frame.message_finished
+                        self._events.append(Pong(payload=frame.payload))
+                    elif frame.opcode is Opcode.CLOSE:
+                        code, reason = frame.payload
+                        if self.state is ConnectionState.LOCAL_CLOSING:
+                            self._state = ConnectionState.CLOSED
+                        else:
+                            self._state = ConnectionState.REMOTE_CLOSING
+                        self._events.append(CloseConnection(code=code, reason=reason))
+                    elif frame.opcode is Opcode.TEXT:
+                        self._events.append(
+                            TextMessage(
+                                data=frame.payload,
+                                frame_finished=frame.frame_finished,
+                                message_finished=frame.message_finished,
+                            )
+                        )
+                    elif frame.opcode is Opcode.BINARY:
+                        self._events.append(
+                            BytesMessage(
+                                data=frame.payload,
+                                frame_finished=frame.frame_finished,
+                                message_finished=frame.message_finished,
+                            )
+                        )
+            except ParseFailed as exc:
+                raise RemoteProtocolError(
+                    str(exc), event_hint=CloseConnection(code=exc.code, reason=str(exc))
```

Comment:
Hmm, I guess if the remote is sending broken data it would probably be wise to send them a close frame then close anyway - assume that the remote would just send more broken data. 
I think we can let the user decide though.

Reply:
This question has come up with the compliance tests, which I've just been patching rather than understanding - I'll understand and fix.